Intermittent nuget restore failures due to invalid package signature validation

### NuGet Product Used

dotnet.exe

### Product Version

5.0.302

### Worked before?

_No response_

### Impact

It's more difficult to complete my work

### Repro Steps & Context

This is [an intermittent issue](https://twitter.com/awakecoding/status/1420380412367491074) that affects CI builds inside Windows containers. Twice this week, a "dotnet publish" command failed to validate Microsoft nuget package signatures while restoring them. This particular build has been working just fine for months before, and once the problem occurs, restarting the build a few times eventually makes it work. For all I know it could be a networking or a timing issue, but I'm not sure what to look for.

The [build script](https://github.com/Devolutions/devolutions-gateway/blob/master/powershell/build.ps1) is for a PowerShell module that contains csproj file used to import dependencies. I have never seen it fail that way outside of the CI environment. As for the CI environment itself, it is Azure Pipelines with our self-hosted build agents running Windows Server 2019, and using Windows Server 2019 build containers.

I don't really expect anyone to reproduce this issue intentionally, but at least if I am not alone then we can start tracking it here. Throwing ideas and theories as to why it may happen sometimes would already be a good start. First, why does it show a lot more information when it fails, and is there something wrong in what is outputted? The validity period for the packages in question is outdated, as it expired on 4/14/2021:

```
Package 'System.Runtime 4.3.0' from source 'https://api.nuget.org/v3/index.json':   SHA256 hash: 0E5F38F57DC1BCC806D8494F4F90FBCEDD988B46760709CBEEC6F4219AA6157D
Package 'System.Runtime 4.3.0' from source 'https://api.nuget.org/v3/index.json':   Valid from: 4/9/2018 8:00:00 PM to 4/14/2021 8:00:00 AM
```

While the validity period is over, code signing usually requires a timestamping server to validate that a package was signed during the validity period of the certificate it was signed with. In other words, it only matters that the package was signed during the validity period, and it will remain valid after that. One idea that comes to mind is an issue with the timestamping server - maybe networking issues could cause this, or maybe the timestamping server used for those packages has issues.

@anangaur @barr

### Verbose Logs

```shell
Here is the log for the failed CI build:

2021-07-28T13:31:33.4657734Z ##[section]Starting: Building PowerShell Module
2021-07-28T13:31:33.4896944Z ==============================================================================
2021-07-28T13:31:33.4897600Z Task         : PowerShell
2021-07-28T13:31:33.4898617Z Description  : Run a PowerShell script on Linux, macOS, or Windows
2021-07-28T13:31:33.4899068Z Version      : 2.189.0
2021-07-28T13:31:33.4899415Z Author       : Microsoft Corporation
2021-07-28T13:31:33.4899902Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/powershell
2021-07-28T13:31:33.4900440Z ==============================================================================
2021-07-28T13:31:37.9744534Z Generating script.
2021-07-28T13:31:37.9795555Z Formatted command: . 'C:\__w\59\s\powershell\build.ps1'
2021-07-28T13:31:38.0094258Z ========================== Starting Command Output ===========================
2021-07-28T13:31:38.0271588Z ##[command]"C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe" -NoLogo -NoProfile -NonInteractive -ExecutionPolicy Unrestricted -Command ". 'C:\__w\_temp\f0e755ae-2018-4c03-a845-d206f5dc7b99.ps1'"
2021-07-28T13:31:40.2806203Z Microsoft (R) Build Engine version 16.10.1+2fd48ab73 for .NET
2021-07-28T13:31:40.2807016Z Copyright (C) Microsoft Corporation. All rights reserved.
2021-07-28T13:31:40.2807199Z 
2021-07-28T13:31:40.6846198Z   Determining projects to restore...
2021-07-28T13:31:42.3356000Z   Package 'System.Runtime 4.3.0' from source 'https://api.nuget.org/v3/index.json': 
2021-07-28T13:31:42.3357702Z   Signature type: Repository
2021-07-28T13:31:42.3359342Z   Package 'Microsoft.NETCore.Platforms 1.1.0' from source 'https://api.nuget.org/v3/index.json': 
2021-07-28T13:31:42.3360997Z   Signature type: Repository
2021-07-28T13:31:42.3362002Z   Package 'System.Collections 4.3.0' from source 'https://api.nuget.org/v3/index.json': 
2021-07-28T13:31:42.3362992Z   Signature type: Repository
2021-07-28T13:31:42.3364361Z   Package 'Microsoft.NETCore.Platforms 1.1.0' from source 'https://api.nuget.org/v3/index.json':   Subject Name: CN=NuGet.org Repository by Microsoft, O=NuGet.org Repository by Microsoft, L=Redmond, S=Washington, C=US
2021-07-28T13:31:42.3368064Z   Package 'System.Runtime 4.3.0' from source 'https://api.nuget.org/v3/index.json':   Subject Name: CN=NuGet.org Repository by Microsoft, O=NuGet.org Repository by Microsoft, L=Redmond, S=Washington, C=US
2021-07-28T13:31:42.3369585Z   Package 'System.Diagnostics.Debug 4.3.0' from source 'https://api.nuget.org/v3/index.json': 
2021-07-28T13:31:42.3370318Z   Signature type: Repository
2021-07-28T13:31:42.3371184Z   Package 'Microsoft.NETCore.Platforms 1.1.0' from source 'https://api.nuget.org/v3/index.json':   SHA256 hash: 0E5F38F57DC1BCC806D8494F4F90FBCEDD988B46760709CBEEC6F4219AA6157D
2021-07-28T13:31:42.3372380Z   Package 'System.Runtime 4.3.0' from source 'https://api.nuget.org/v3/index.json':   SHA256 hash: 0E5F38F57DC1BCC806D8494F4F90FBCEDD988B46760709CBEEC6F4219AA6157D
2021-07-28T13:31:42.3373731Z   Package 'System.Runtime 4.3.0' from source 'https://api.nuget.org/v3/index.json':   Valid from: 4/9/2018 8:00:00 PM to 4/14/2021 8:00:00 AM
2021-07-28T13:31:42.3374756Z   Package 'System.Threading.Tasks 4.3.0' from source 'https://api.nuget.org/v3/index.json': 
2021-07-28T13:31:42.3375619Z   Signature type: Repository
2021-07-28T13:31:42.3376974Z   Package 'System.Diagnostics.Debug 4.3.0' from source 'https://api.nuget.org/v3/index.json':   Subject Name: CN=NuGet.org Repository by Microsoft, O=NuGet.org Repository by Microsoft, L=Redmond, S=Washington, C=US
2021-07-28T13:31:42.3381840Z   Package 'System.Diagnostics.Debug 4.3.0' from source 'https://api.nuget.org/v3/index.json':   SHA256 hash: 0E5F38F57DC1BCC806D8494F4F90FBCEDD988B46760709CBEEC6F4219AA6157D
2021-07-28T13:31:42.3383505Z   Package 'System.Diagnostics.Debug 4.3.0' from source 'https://api.nuget.org/v3/index.json':   Valid from: 4/9/2018 8:00:00 PM to 4/14/2021 8:00:00 AM
2021-07-28T13:31:42.3384747Z   Package 'System.Globalization 4.3.0' from source 'https://api.nuget.org/v3/index.json': 
2021-07-28T13:31:42.3387992Z   Signature type: Repository
2021-07-28T13:31:42.3389810Z   Package 'System.Threading.Tasks 4.3.0' from source 'https://api.nuget.org/v3/index.json':   Subject Name: CN=NuGet.org Repository by Microsoft, O=NuGet.org Repository by Microsoft, L=Redmond, S=Washington, C=US
2021-07-28T13:31:42.3392494Z   Package 'System.Collections 4.3.0' from source 'https://api.nuget.org/v3/index.json':   Subject Name: CN=NuGet.org Repository by Microsoft, O=NuGet.org Repository by Microsoft, L=Redmond, S=Washington, C=US
2021-07-28T13:31:42.3394303Z   Package 'System.Threading.Tasks 4.3.0' from source 'https://api.nuget.org/v3/index.json':   SHA256 hash: 0E5F38F57DC1BCC806D8494F4F90FBCEDD988B46760709CBEEC6F4219AA6157D
2021-07-28T13:31:42.3395894Z   Package 'System.Threading.Tasks 4.3.0' from source 'https://api.nuget.org/v3/index.json':   Valid from: 4/9/2018 8:00:00 PM to 4/14/2021 8:00:00 AM
2021-07-28T13:31:42.3397476Z   Package 'System.Globalization 4.3.0' from source 'https://api.nuget.org/v3/index.json':   Subject Name: CN=NuGet.org Repository by Microsoft, O=NuGet.org Repository by Microsoft, L=Redmond, S=Washington, C=US
2021-07-28T13:31:42.3399139Z   Package 'System.Collections 4.3.0' from source 'https://api.nuget.org/v3/index.json':   SHA256 hash: 0E5F38F57DC1BCC806D8494F4F90FBCEDD988B46760709CBEEC6F4219AA6157D
2021-07-28T13:31:42.3400649Z   Package 'Microsoft.NETCore.Platforms 1.1.0' from source 'https://api.nuget.org/v3/index.json':   Valid from: 4/9/2018 8:00:00 PM to 4/14/2021 8:00:00 AM
2021-07-28T13:31:42.3402209Z   Package 'System.Globalization 4.3.0' from source 'https://api.nuget.org/v3/index.json':   SHA256 hash: 0E5F38F57DC1BCC806D8494F4F90FBCEDD988B46760709CBEEC6F4219AA6157D
2021-07-28T13:31:42.3403858Z   Package 'System.Collections 4.3.0' from source 'https://api.nuget.org/v3/index.json':   Valid from: 4/9/2018 8:00:00 PM to 4/14/2021 8:00:00 AM
2021-07-28T13:31:42.3405366Z   Package 'System.Globalization 4.3.0' from source 'https://api.nuget.org/v3/index.json':   Valid from: 4/9/2018 8:00:00 PM to 4/14/2021 8:00:00 AM
2021-07-28T13:31:42.3416116Z C:\__w\59\s\powershell\DevolutionsGateway\src\DevolutionsGateway.csproj : error NU3037: Package 'System.Collections 4.3.0' from source 'https://api.nuget.org/v3/index.json': The repository primary signature validity period has expired. [C:\__w\59\s\powershell\DevolutionsGateway\src\DevolutionsGateway.sln]
2021-07-28T13:31:42.3419040Z C:\__w\59\s\powershell\DevolutionsGateway\src\DevolutionsGateway.csproj : error NU3037: Package 'System.Runtime 4.3.0' from source 'https://api.nuget.org/v3/index.json': The repository primary signature validity period has expired. [C:\__w\59\s\powershell\DevolutionsGateway\src\DevolutionsGateway.sln]
2021-07-28T13:31:42.3421591Z C:\__w\59\s\powershell\DevolutionsGateway\src\DevolutionsGateway.csproj : error NU3037: Package 'System.Globalization 4.3.0' from source 'https://api.nuget.org/v3/index.json': The repository primary signature validity period has expired. [C:\__w\59\s\powershell\DevolutionsGateway\src\DevolutionsGateway.sln]
2021-07-28T13:31:42.3424744Z C:\__w\59\s\powershell\DevolutionsGateway\src\DevolutionsGateway.csproj : error NU3028: Package 'System.Collections 4.3.0' from source 'https://api.nuget.org/v3/index.json': The repository primary signature's timestamping certificate is not trusted by the trust provider. [C:\__w\59\s\powershell\DevolutionsGateway\src\DevolutionsGateway.sln]
2021-07-28T13:31:42.3427366Z C:\__w\59\s\powershell\DevolutionsGateway\src\DevolutionsGateway.csproj : error NU3028: Package 'System.Globalization 4.3.0' from source 'https://api.nuget.org/v3/index.json': The repository primary signature's timestamping certificate is not trusted by the trust provider. [C:\__w\59\s\powershell\DevolutionsGateway\src\DevolutionsGateway.sln]
2021-07-28T13:31:42.3429471Z C:\__w\59\s\powershell\DevolutionsGateway\src\DevolutionsGateway.csproj : error NU3028: Package 'System.Runtime 4.3.0' from source 'https://api.nuget.org/v3/index.json': The repository primary signature's timestamping certificate is not trusted by the trust provider. [C:\__w\59\s\powershell\DevolutionsGateway\src\DevolutionsGateway.sln]
2021-07-28T13:31:42.3433492Z C:\__w\59\s\powershell\DevolutionsGateway\src\DevolutionsGateway.csproj : error NU3037: Package 'System.Threading.Tasks 4.3.0' from source 'https://api.nuget.org/v3/index.json': The repository primary signature validity period has expired. [C:\__w\59\s\powershell\DevolutionsGateway\src\DevolutionsGateway.sln]
2021-07-28T13:31:42.3436512Z C:\__w\59\s\powershell\DevolutionsGateway\src\DevolutionsGateway.csproj : error NU3037: Package 'System.Diagnostics.Debug 4.3.0' from source 'https://api.nuget.org/v3/index.json': The repository primary signature validity period has expired. [C:\__w\59\s\powershell\DevolutionsGateway\src\DevolutionsGateway.sln]
2021-07-28T13:31:42.3439531Z C:\__w\59\s\powershell\DevolutionsGateway\src\DevolutionsGateway.csproj : error NU3037: Package 'Microsoft.NETCore.Platforms 1.1.0' from source 'https://api.nuget.org/v3/index.json': The repository primary signature validity period has expired. [C:\__w\59\s\powershell\DevolutionsGateway\src\DevolutionsGateway.sln]
2021-07-28T13:31:42.3442252Z C:\__w\59\s\powershell\DevolutionsGateway\src\DevolutionsGateway.csproj : error NU3028: Package 'System.Threading.Tasks 4.3.0' from source 'https://api.nuget.org/v3/index.json': The repository primary signature's timestamping certificate is not trusted by the trust provider. [C:\__w\59\s\powershell\DevolutionsGateway\src\DevolutionsGateway.sln]
2021-07-28T13:31:42.3444760Z C:\__w\59\s\powershell\DevolutionsGateway\src\DevolutionsGateway.csproj : error NU3028: Package 'System.Diagnostics.Debug 4.3.0' from source 'https://api.nuget.org/v3/index.json': The repository primary signature's timestamping certificate is not trusted by the trust provider. [C:\__w\59\s\powershell\DevolutionsGateway\src\DevolutionsGateway.sln]
2021-07-28T13:31:42.3446740Z C:\__w\59\s\powershell\DevolutionsGateway\src\DevolutionsGateway.csproj : error NU3028: Package 'Microsoft.NETCore.Platforms 1.1.0' from source 'https://api.nuget.org/v3/index.json': The repository primary signature's timestamping certificate is not trusted by the trust provider. [C:\__w\59\s\powershell\DevolutionsGateway\src\DevolutionsGateway.sln]
2021-07-28T13:31:42.4788849Z   Failed to restore C:\__w\59\s\powershell\DevolutionsGateway\src\DevolutionsGateway.csproj (in 1.44 sec).
2021-07-28T13:31:42.7421339Z Copy-Item : Cannot find path 'C:\__w\59\s\powershell\DevolutionsGateway\bin' 
2021-07-28T13:31:42.7425532Z because it does not exist.
2021-07-28T13:31:42.7426271Z At C:\__w\59\s\powershell\build.ps1:22 char:1
2021-07-28T13:31:42.7426897Z + Copy-Item "$PSScriptRoot\$ModuleName\bin" -Destination "$PSModuleOutp ...
2021-07-28T13:31:42.7427534Z + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2021-07-28T13:31:42.7428185Z     + CategoryInfo          : ObjectNotFound: (C:\__w\59\s\pow...ionsGateway\b 
2021-07-28T13:31:42.7429217Z    in:String) [Copy-Item], ItemNotFoundException
2021-07-28T13:31:42.7429652Z     + FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.CopyI 
2021-07-28T13:31:42.7429968Z    temCommand
2021-07-28T13:31:42.7430184Z  
2021-07-28T13:31:42.8271794Z ##[error]PowerShell exited with code '1'.
2021-07-28T13:31:42.8604817Z ##[section]Finishing: Building PowerShell Module

And here is the log for the exact same build job that worked the second time:

2021-07-28T15:29:07.3913232Z ##[section]Starting: Building PowerShell Module
2021-07-28T15:29:07.4169381Z ==============================================================================
2021-07-28T15:29:07.4170211Z Task         : PowerShell
2021-07-28T15:29:07.4170486Z Description  : Run a PowerShell script on Linux, macOS, or Windows
2021-07-28T15:29:07.4170742Z Version      : 2.189.0
2021-07-28T15:29:07.4170969Z Author       : Microsoft Corporation
2021-07-28T15:29:07.4171266Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/powershell
2021-07-28T15:29:07.4171595Z ==============================================================================
2021-07-28T15:29:11.9329867Z Generating script.
2021-07-28T15:29:11.9372617Z Formatted command: . 'C:\__w\59\s\powershell\build.ps1'
2021-07-28T15:29:11.9673045Z ========================== Starting Command Output ===========================
2021-07-28T15:29:11.9857095Z ##[command]"C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe" -NoLogo -NoProfile -NonInteractive -ExecutionPolicy Unrestricted -Command ". 'C:\__w\_temp\4e16f1cc-5a6e-4de9-b9e3-f223db604452.ps1'"
2021-07-28T15:29:14.3315700Z Microsoft (R) Build Engine version 16.10.2+857e5a733 for .NET
2021-07-28T15:29:14.3316493Z Copyright (C) Microsoft Corporation. All rights reserved.
2021-07-28T15:29:14.3316676Z 
2021-07-28T15:29:14.7371296Z   Determining projects to restore...
2021-07-28T15:29:17.3209204Z   Restored C:\__w\59\s\powershell\DevolutionsGateway\src\DevolutionsGateway.csproj (in 2.27 sec).
2021-07-28T15:29:18.9191614Z   DevolutionsGateway -> C:\__w\59\s\powershell\DevolutionsGateway\src\bin\Release\netstandard2.0\DevolutionsGateway.dll
2021-07-28T15:29:18.9456837Z   DevolutionsGateway -> C:\__w\59\s\powershell\DevolutionsGateway\bin\
2021-07-28T15:29:19.2000312Z ##[section]Finishing: Building PowerShell Module

And for reference, here is the Dockerfile for the build container I use:

ARG FROM_IMAGE=mcr.microsoft.com/powershell:windowsservercore-1809
FROM ${FROM_IMAGE}

SHELL ["cmd", "/S", "/C"]

# Download channel for fixed install.
RUN mkdir "C:\TEMP"
ARG CHANNEL_URL=https://aka.ms/vs/16/release/channel
ADD ${CHANNEL_URL} "C:\\TEMP\\VisualStudio.chman"

# Download the Build Tools bootstrapper.
ADD https://aka.ms/vs/16/release/vs_buildtools.exe "C:\TEMP\vs_buildtools.exe"

# Install Build Tools excluding workloads and components with known issues.
RUN C:\TEMP\vs_buildtools.exe --quiet --wait --norestart --nocache \
    --installPath C:\BuildTools \
    --channelUri C:\TEMP\VisualStudio.chman \
    --installChannelUri C:\TEMP\VisualStudio.chman \
    --includeRecommended \
    --add Microsoft.Component.MSBuild \
    --add Microsoft.Component.VC.Runtime.UCRTSDK \
    --add Microsoft.VisualStudio.Workload.VCTools \
    --add Microsoft.VisualStudio.Component.VC.140 \
    --add Microsoft.VisualStudio.Component.VC.v141.x86.x64 \
    --add Microsoft.VisualStudio.Component.VC.v141.ARM64 \
    --add Microsoft.VisualStudio.Component.VC.Tools.x86.x64 \
    --add Microsoft.VisualStudio.Component.VC.Tools.ARM64 \
    --add Microsoft.VisualStudio.Component.Windows10SDK.17763 \
    --add Microsoft.VisualStudio.Workload.ManagedDesktopBuildTools \
    --add Microsoft.VisualStudio.Workload.XamarinBuildTools \
    --add Microsoft.Net.Component.4.7.2.TargetingPack \
    --remove Microsoft.VisualStudio.Component.Windows10SDK.10240 \
    --remove Microsoft.VisualStudio.Component.Windows10SDK.10586 \
    --remove Microsoft.VisualStudio.Component.Windows10SDK.14393 \
	--remove Microsoft.VisualStudio.Component.Windows81SDK \
 || IF "%ERRORLEVEL%"=="3010" EXIT 0

SHELL ["powershell.exe", "-ExecutionPolicy", "Bypass", "-Command"]
ENV chocolateyUseWindowsCompression=false

RUN Set-Service -Name wuauserv -StartupType Manual; Install-WindowsFeature -Name NET-Framework-Features -Verbose

RUN iex ((New-Object System.Net.WebClient).DownloadString('https://chocolatey.org/install.ps1'))
RUN choco install --no-progress --yes git --params '"/InstallDir:C:\tools\git"'; \
    choco install --no-progress --yes git-lfs; \
    choco install --no-progress --yes cmake --version=3.16.2 --params '"/InstallDir:C:\tools\cmake"' --installargs 'ADD_CMAKE_TO_PATH=""System""'; \
    choco install --no-progress --yes ninja ; \
    choco install --no-progress --yes python ; \
    choco install --no-progress --yes 7zip ; \
    choco install --no-progress --yes yasm ; \
    choco install --no-progress --yes vswhere ; \
    choco install --no-progress --yes wixtoolset --version 3.11.2 ; \
    choco install --no-progress --yes nuget.commandline ; \
    choco install --no-progress --yes openjdk11

RUN New-ItemProperty -Path "HKLM:\SOFTWARE\Microsoft\.NETFramework\v4.0.30319" -Name "SchUseStrongCrypto" -Value "1" -Type DWORD
RUN New-ItemProperty -Path "HKLM:\SOFTWARE\Wow6432Node\Microsoft\.NETFramework\v4.0.30319" -Name "SchUseStrongCrypto" -Value "1" -Type DWORD
RUN Install-PackageProvider -Name NuGet -Force
RUN Install-Module –Name PowerShellGet –Force

# Rust
ARG RUST_VERSION='1.52.0'
ADD https://static.rust-lang.org/rustup/dist/x86_64-pc-windows-msvc/rustup-init.exe "C:\TEMP\rustup-init.exe"
RUN C:\TEMP\rustup-init.exe -y
RUN rustup toolchain install ${Env:RUST_VERSION}-x86_64-pc-windows-msvc
RUN rustup default ${Env:RUST_VERSION}-x86_64-pc-windows-msvc
RUN rustup target install i686-pc-windows-msvc
RUN rustup target install x86_64-pc-windows-msvc
RUN rustup target install aarch64-pc-windows-msvc

SHELL ["cmd", "/S", "/C"]

# Remove temporary files
RUN rmdir /S /Q C:\TEMP

RUN setx /m path "%PATH%;c:\Program Files (x86)\WiX Toolset v3.11\bin;C:\BuildTools\MSBuild\Current\Bin;c:\Program Files (x86)\Windows Kits\10\App Certification Kit;c:\BuildTools\Common7\IDE;c:\Program Files (x86)\Windows Kits\10\bin\10.0.17763.0\x64"

# Default to PowerShell if no other command specified.
CMD ["pwsh.exe", "-NoLogo", "-ExecutionPolicy", "Bypass"]
```

--- a/src/NuGet.Core/NuGet.Packaging.Extraction/NuGet.Packaging.Extraction.csproj
+++ b/src/NuGet.Core/NuGet.Packaging.Extraction/NuGet.Packaging.Extraction.csproj
@@ -177,6 +177,10 @@
     <Compile Include="..\NuGet.Packaging\Signing\Archive\Crc32.cs" Link="NuGet.Packaging\Signing\Archive\Crc32.cs" />
     <Compile Include="..\NuGet.Packaging\Signing\Archive\UnsignedPackageArchiveMetadata.cs" Link="NuGet.Packaging\Signing\Archive\UnsignedPackageArchiveMetadata.cs" />
     <Compile Include="..\NuGet.Packaging\Signing\Archive\SignedPackageArchiveMetadata.cs" Link="NuGet.Packaging\Signing\Archive\SignedPackageArchiveMetadata.cs" />
+    <Compile Include="..\NuGet.Packaging\Signing\ChainBuilding\DefaultX509ChainBuildPolicy.cs" Link="NuGet.Packaging\Signing\ChainBuilding\DefaultX509ChainBuildPolicy.cs" />
+    <Compile Include="..\NuGet.Packaging\Signing\ChainBuilding\IX509ChainBuildPolicy.cs" Link="NuGet.Packaging\Signing\ChainBuilding\IX509ChainBuildPolicy.cs" />
+    <Compile Include="..\NuGet.Packaging\Signing\ChainBuilding\RetriableX509ChainBuildPolicy.cs" Link="NuGet.Packaging\Signing\ChainBuilding\RetriableX509ChainBuildPolicy.cs" />
+    <Compile Include="..\NuGet.Packaging\Signing\ChainBuilding\X509ChainBuildPolicyFactory.cs" Link="NuGet.Packaging\Signing\ChainBuilding\X509ChainBuildPolicyFactory.cs" />
     <Compile Include="..\NuGet.Packaging\Signing\Cms\CmsFactory.cs" Link="NuGet.Packaging\Signing\Cms\CmsFactory.cs" />
     <Compile Include="..\NuGet.Packaging\Signing\Cms\ICms.cs" Link="NuGet.Packaging\Signing\Cms\ICms.cs" />
     <Compile Include="..\NuGet.Packaging\Signing\Cms\NativeCms.cs" Link="NuGet.Packaging\Signing\Cms\NativeCms.cs" />

--- a/src/NuGet.Core/NuGet.Packaging/Signing/ChainBuilding/DefaultX509ChainBuildPolicy.cs
+++ b/src/NuGet.Core/NuGet.Packaging/Signing/ChainBuilding/DefaultX509ChainBuildPolicy.cs
@@ -1,0 +1,28 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Security.Cryptography.X509Certificates;
+
+namespace NuGet.Packaging.Signing
+{
+    internal sealed class DefaultX509ChainBuildPolicy : IX509ChainBuildPolicy
+    {
+        internal static IX509ChainBuildPolicy Instance { get; } = new DefaultX509ChainBuildPolicy();
+
+        public bool Build(X509Chain chain, X509Certificate2 certificate)
+        {
+            if (chain is null)
+            {
+                throw new ArgumentNullException(nameof(chain));
+            }
+
+            if (certificate is null)
+            {
+                throw new ArgumentNullException(nameof(certificate));
+            }
+
+            return chain.Build(certificate);
+        }
+    }
+}

--- a/src/NuGet.Core/NuGet.Packaging/Signing/ChainBuilding/DefaultX509ChainBuildPolicy.cs
+++ b/src/NuGet.Core/NuGet.Packaging/Signing/ChainBuilding/DefaultX509ChainBuildPolicy.cs
@@ -10,6 +10,8 @@ namespace NuGet.Packaging.Signing
     {
         internal static IX509ChainBuildPolicy Instance { get; } = new DefaultX509ChainBuildPolicy();
 
+        private DefaultX509ChainBuildPolicy() { }
+
         public bool Build(X509Chain chain, X509Certificate2 certificate)
         {
             if (chain is null)

--- a/src/NuGet.Core/NuGet.Packaging/Signing/ChainBuilding/IX509ChainBuildPolicy.cs
+++ b/src/NuGet.Core/NuGet.Packaging/Signing/ChainBuilding/IX509ChainBuildPolicy.cs
@@ -1,0 +1,16 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Security.Cryptography.X509Certificates;
+
+namespace NuGet.Packaging.Signing
+{
+    /// <summary>
+    /// Represents a wrapper around <see cref="X509Chain.Build(X509Certificate2)" /> to enable
+    /// custom behaviors (e.g.:  retry on failure).
+    /// </summary>
+    internal interface IX509ChainBuildPolicy
+    {
+        bool Build(X509Chain chain, X509Certificate2 certificate);
+    }
+}

--- a/src/NuGet.Core/NuGet.Packaging/Signing/ChainBuilding/RetriableX509ChainBuildPolicy.cs
+++ b/src/NuGet.Core/NuGet.Packaging/Signing/ChainBuilding/RetriableX509ChainBuildPolicy.cs
@@ -10,11 +10,17 @@ namespace NuGet.Packaging.Signing
     internal sealed class RetriableX509ChainBuildPolicy : IX509ChainBuildPolicy
     {
         // These properties are non-private only to facilitate testing.
+        internal IX509ChainBuildPolicy InnerPolicy { get; }
         internal int RetryCount { get; }
         internal TimeSpan SleepInterval { get; }
 
-        internal RetriableX509ChainBuildPolicy(int retryCount, TimeSpan sleepInterval)
+        internal RetriableX509ChainBuildPolicy(IX509ChainBuildPolicy innerPolicy, int retryCount, TimeSpan sleepInterval)
         {
+            if (innerPolicy is null)
+            {
+                throw new ArgumentNullException(nameof(innerPolicy));
+            }
+
             if (retryCount < 1)
             {
                 throw new ArgumentOutOfRangeException(nameof(retryCount));
@@ -25,6 +31,7 @@ namespace NuGet.Packaging.Signing
                 throw new ArgumentOutOfRangeException(nameof(sleepInterval));
             }
 
+            InnerPolicy = innerPolicy;
             RetryCount = retryCount;
             SleepInterval = sleepInterval;
         }
@@ -41,7 +48,7 @@ namespace NuGet.Packaging.Signing
                 throw new ArgumentNullException(nameof(certificate));
             }
 
-            bool wasBuilt = chain.Build(certificate);
+            bool wasBuilt = InnerPolicy.Build(chain, certificate);
 
             for (var i = 0; i < RetryCount; ++i)
             {
@@ -68,7 +75,7 @@ namespace NuGet.Packaging.Signing
 
                 Thread.Sleep(SleepInterval);
 
-                wasBuilt = chain.Build(certificate);
+                wasBuilt = InnerPolicy.Build(chain, certificate);
             }
 
             return wasBuilt;

--- a/src/NuGet.Core/NuGet.Packaging/Signing/ChainBuilding/RetriableX509ChainBuildPolicy.cs
+++ b/src/NuGet.Core/NuGet.Packaging/Signing/ChainBuilding/RetriableX509ChainBuildPolicy.cs
@@ -1,0 +1,77 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading;
+
+namespace NuGet.Packaging.Signing
+{
+    internal sealed class RetriableX509ChainBuildPolicy : IX509ChainBuildPolicy
+    {
+        // These properties are non-private only to facilitate testing.
+        internal int RetryCount { get; }
+        internal TimeSpan SleepInterval { get; }
+
+        internal RetriableX509ChainBuildPolicy(int retryCount, TimeSpan sleepInterval)
+        {
+            if (retryCount < 1)
+            {
+                throw new ArgumentOutOfRangeException(nameof(retryCount));
+            }
+
+            if (sleepInterval < TimeSpan.Zero)
+            {
+                throw new ArgumentOutOfRangeException(nameof(sleepInterval));
+            }
+
+            RetryCount = retryCount;
+            SleepInterval = sleepInterval;
+        }
+
+        public bool Build(X509Chain chain, X509Certificate2 certificate)
+        {
+            if (chain is null)
+            {
+                throw new ArgumentNullException(nameof(chain));
+            }
+
+            if (certificate is null)
+            {
+                throw new ArgumentNullException(nameof(certificate));
+            }
+
+            bool wasBuilt = chain.Build(certificate);
+
+            for (var i = 0; i < RetryCount; ++i)
+            {
+                if (wasBuilt)
+                {
+                    break;
+                }
+
+                bool hasUntrustedRoot = false;
+
+                foreach (X509ChainStatus chainStatus in chain.ChainStatus)
+                {
+                    if (chainStatus.Status.HasFlag(X509ChainStatusFlags.UntrustedRoot))
+                    {
+                        hasUntrustedRoot = true;
+                        break;
+                    }
+                }
+
+                if (!hasUntrustedRoot)
+                {
+                    break;
+                }
+
+                Thread.Sleep(SleepInterval);
+
+                wasBuilt = chain.Build(certificate);
+            }
+
+            return wasBuilt;
+        }
+    }
+}

--- a/src/NuGet.Core/NuGet.Packaging/Signing/ChainBuilding/X509ChainBuildPolicyFactory.cs
+++ b/src/NuGet.Core/NuGet.Packaging/Signing/ChainBuilding/X509ChainBuildPolicyFactory.cs
@@ -12,7 +12,31 @@ namespace NuGet.Packaging.Signing
         internal const string EnvironmentVariableName = "NUGET_EXPERIMENTAL_CHAIN_BUILD_RETRY_POLICY";
         internal const char ValueDelimiter = ',';
 
+        private static readonly object LockObject = new object();
+        private static IX509ChainBuildPolicy Policy;
+
         internal static IX509ChainBuildPolicy Create(IEnvironmentVariableReader reader)
+        {
+            if (reader is null)
+            {
+                throw new ArgumentNullException(nameof(reader));
+            }
+
+            if (Policy is object)
+            {
+                return Policy;
+            }
+
+            lock (LockObject)
+            {
+                Policy ??= CreateWithoutCaching(reader);
+            }
+
+            return Policy;
+        }
+
+        // This is non-private only to facilitate testing.
+        internal static IX509ChainBuildPolicy CreateWithoutCaching(IEnvironmentVariableReader reader)
         {
             if (reader is null)
             {
@@ -38,7 +62,7 @@ namespace NuGet.Packaging.Signing
                 {
                     TimeSpan sleepInterval = TimeSpan.FromMilliseconds(sleepIntervalInMilliseconds);
 
-                    return new RetriableX509ChainBuildPolicy(retryCount, sleepInterval);
+                    return new RetriableX509ChainBuildPolicy(DefaultX509ChainBuildPolicy.Instance, retryCount, sleepInterval);
                 }
             }
 

--- a/src/NuGet.Core/NuGet.Packaging/Signing/ChainBuilding/X509ChainBuildPolicyFactory.cs
+++ b/src/NuGet.Core/NuGet.Packaging/Signing/ChainBuilding/X509ChainBuildPolicyFactory.cs
@@ -1,0 +1,48 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using NuGet.Common;
+
+namespace NuGet.Packaging.Signing
+{
+    internal static class X509ChainBuildPolicyFactory
+    {
+        // These fields are non-private only to facilitate testing.
+        internal const string EnvironmentVariableName = "NUGET_EXPERIMENTAL_CHAIN_BUILD_RETRY_POLICY";
+        internal const char ValueDelimiter = ',';
+
+        internal static IX509ChainBuildPolicy Create(IEnvironmentVariableReader reader)
+        {
+            if (reader is null)
+            {
+                throw new ArgumentNullException(nameof(reader));
+            }
+
+            if (RuntimeEnvironmentHelper.IsWindows)
+            {
+                string value = reader.GetEnvironmentVariable(EnvironmentVariableName);
+
+                if (string.IsNullOrWhiteSpace(value))
+                {
+                    return DefaultX509ChainBuildPolicy.Instance;
+                }
+
+                string[] parts = value.Split(ValueDelimiter);
+
+                if (parts.Length == 2
+                    && int.TryParse(parts[0], out int retryCount)
+                    && retryCount > 0
+                    && int.TryParse(parts[1], out int sleepIntervalInMilliseconds)
+                    && sleepIntervalInMilliseconds >= 0)
+                {
+                    TimeSpan sleepInterval = TimeSpan.FromMilliseconds(sleepIntervalInMilliseconds);
+
+                    return new RetriableX509ChainBuildPolicy(retryCount, sleepInterval);
+                }
+            }
+
+            return DefaultX509ChainBuildPolicy.Instance;
+        }
+    }
+}

--- a/src/NuGet.Core/NuGet.Packaging/Signing/Utility/CertificateChainUtility.cs
+++ b/src/NuGet.Core/NuGet.Packaging/Signing/Utility/CertificateChainUtility.cs
@@ -65,7 +65,7 @@ namespace NuGet.Packaging.Signing
 
                 chain.ChainPolicy.RevocationMode = X509RevocationMode.Online;
 
-                if (chain.Build(certificate))
+                if (BuildWithPolicy(chain, certificate))
                 {
                     return GetCertificateChain(chain);
                 }
@@ -184,12 +184,29 @@ namespace NuGet.Packaging.Signing
                 throw new ArgumentNullException(nameof(certificate));
             }
 
-            var buildSuccess = chain.Build(certificate);
+            bool buildSuccess = BuildWithPolicy(chain, certificate);
             status = new X509ChainStatus[chain.ChainStatus.Length];
-            chain.ChainStatus.CopyTo(status, 0);
+            chain.ChainStatus.CopyTo(status, index: 0);
 
             // Check if time is not in the future
             return buildSuccess && !CertificateUtility.IsCertificateValidityPeriodInTheFuture(certificate);
+        }
+
+        internal static bool BuildWithPolicy(X509Chain chain, X509Certificate2 certificate)
+        {
+            if (chain is null)
+            {
+                throw new ArgumentNullException(nameof(chain));
+            }
+
+            if (certificate is null)
+            {
+                throw new ArgumentNullException(nameof(certificate));
+            }
+
+            IX509ChainBuildPolicy policy = X509ChainBuildPolicyFactory.Create(EnvironmentVariableWrapper.Instance);
+
+            return policy.Build(chain, certificate);
         }
 
         // Ignore some chain status flags to special case them

--- a/src/NuGet.Core/NuGet.Packaging/Signing/Utility/CertificateUtility.cs
+++ b/src/NuGet.Core/NuGet.Packaging/Signing/Utility/CertificateUtility.cs
@@ -14,7 +14,7 @@ namespace NuGet.Packaging.Signing
 {
     public static class CertificateUtility
     {
-        private const int _limit = 10;
+        private const int ChainDepthLimit = 10;
 
         /// <summary>
         /// Converts a X509Certificate2 to a human friendly string of the following format -
@@ -91,16 +91,16 @@ namespace NuGet.Packaging.Signing
 
             collectionStringBuilder.AppendLine(Strings.CertUtilityMultipleCertificatesHeader);
 
-            for (var i = 0; i < Math.Min(_limit, certCollection.Count); i++)
+            for (var i = 0; i < Math.Min(ChainDepthLimit, certCollection.Count); i++)
             {
                 var cert = certCollection[i];
                 X509Certificate2ToString(cert, collectionStringBuilder, fingerprintAlgorithm, indentation: "  ");
                 collectionStringBuilder.AppendLine();
             }
 
-            if (certCollection.Count > _limit)
+            if (certCollection.Count > ChainDepthLimit)
             {
-                collectionStringBuilder.AppendLine(string.Format(Strings.CertUtilityMultipleCertificatesFooter, certCollection.Count - _limit));
+                collectionStringBuilder.AppendLine(string.Format(Strings.CertUtilityMultipleCertificatesFooter, certCollection.Count - ChainDepthLimit));
             }
 
             return collectionStringBuilder.ToString();
@@ -114,15 +114,15 @@ namespace NuGet.Packaging.Signing
 
             var chainElementsCount = chain.ChainElements.Count;
             // Start in 1 to omit main certificate (only build the chain)
-            for (var i = 1; i < Math.Min(_limit, chainElementsCount); i++)
+            for (var i = 1; i < Math.Min(ChainDepthLimit, chainElementsCount); i++)
             {
                 X509Certificate2ToString(chain.ChainElements[i].Certificate, collectionStringBuilder, fingerprintAlgorithm, indentation);
                 indentation += indentationLevel;
             }
 
-            if (chainElementsCount > _limit)
+            if (chainElementsCount > ChainDepthLimit)
             {
-                collectionStringBuilder.AppendLine(string.Format(Strings.CertUtilityMultipleCertificatesFooter, chainElementsCount - _limit));
+                collectionStringBuilder.AppendLine(string.Format(Strings.CertUtilityMultipleCertificatesFooter, chainElementsCount - ChainDepthLimit));
             }
 
             return collectionStringBuilder.ToString();
@@ -317,7 +317,7 @@ namespace NuGet.Packaging.Signing
 
             using (var chainHolder = new X509ChainHolder())
             {
-                var chain = chainHolder.Chain;
+                X509Chain chain = chainHolder.Chain;
 
                 chain.ChainPolicy.RevocationMode = X509RevocationMode.NoCheck;
                 chain.ChainPolicy.VerificationFlags = X509VerificationFlags.AllowUnknownCertificateAuthority |
@@ -325,7 +325,7 @@ namespace NuGet.Packaging.Signing
                     X509VerificationFlags.IgnoreCertificateAuthorityRevocationUnknown |
                     X509VerificationFlags.IgnoreEndRevocationUnknown;
 
-                chain.Build(certificate);
+                CertificateChainUtility.BuildWithPolicy(chain, certificate);
 
                 if (chain.ChainElements.Count != 1)
                 {

--- a/src/NuGet.Core/NuGet.Packaging/Signing/Utility/SignatureUtility.cs
+++ b/src/NuGet.Core/NuGet.Packaging/Signing/Utility/SignatureUtility.cs
@@ -619,13 +619,13 @@ namespace NuGet.Packaging.Signing
 
             using (var chainHolder = new X509ChainHolder())
             {
-                var chain = chainHolder.Chain;
+                X509Chain chain = chainHolder.Chain;
 
                 chain.ChainPolicy.ExtraStore.AddRange(extraStore);
 
                 chain.ChainPolicy.RevocationMode = X509RevocationMode.NoCheck;
 
-                chain.Build(certificate);
+                CertificateChainUtility.BuildWithPolicy(chain, certificate);
 
                 if (chain.ChainStatus.Any(chainStatus =>
                     chainStatus.Status.HasFlag(X509ChainStatusFlags.Cyclic) ||

--- a/test/NuGet.Core.Tests/NuGet.Packaging.Test/SigningTests/CertificateChainUtilityTests.cs
+++ b/test/NuGet.Core.Tests/NuGet.Packaging.Test/SigningTests/CertificateChainUtilityTests.cs
@@ -188,5 +188,29 @@ namespace NuGet.Packaging.Test
                 }
             }
         }
+
+        [Fact]
+        public void BuildWithPolicy_WhenChainIsNull_Throws()
+        {
+            using (X509Certificate2 certificate = _fixture.GetDefaultCertificate())
+            {
+                ArgumentNullException exception = Assert.Throws<ArgumentNullException>(
+                    () => CertificateChainUtility.BuildWithPolicy(chain: null, certificate));
+
+                Assert.Equal("chain", exception.ParamName);
+            }
+        }
+
+        [Fact]
+        public void BuildWithPolicy_WhenCertificateIsNull_Throws()
+        {
+            using (var chain = new X509Chain())
+            {
+                ArgumentNullException exception = Assert.Throws<ArgumentNullException>(
+                    () => CertificateChainUtility.BuildWithPolicy(chain, certificate: null));
+
+                Assert.Equal("certificate", exception.ParamName);
+            }
+        }
     }
 }

--- a/test/NuGet.Core.Tests/NuGet.Packaging.Test/SigningTests/ChainBuilding/DefaultX509ChainBuildPolicyTests.cs
+++ b/test/NuGet.Core.Tests/NuGet.Packaging.Test/SigningTests/ChainBuilding/DefaultX509ChainBuildPolicyTests.cs
@@ -1,0 +1,66 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Security.Cryptography.X509Certificates;
+using NuGet.Packaging.Signing;
+using Xunit;
+
+namespace NuGet.Packaging.Test
+{
+    public class DefaultX509ChainBuildPolicyTests : IClassFixture<CertificatesFixture>
+    {
+        private readonly CertificatesFixture _fixture;
+
+        public DefaultX509ChainBuildPolicyTests(CertificatesFixture fixture)
+        {
+            _fixture = fixture ?? throw new ArgumentNullException(nameof(fixture));
+        }
+
+        [Fact]
+        public void Instance_Always_IsIdempotent()
+        {
+            IX509ChainBuildPolicy instance0 = DefaultX509ChainBuildPolicy.Instance;
+            IX509ChainBuildPolicy instance1 = DefaultX509ChainBuildPolicy.Instance;
+
+            Assert.Same(instance0, instance1);
+        }
+
+        [Fact]
+        public void Build_WhenChainIsNull_Throws()
+        {
+            using (X509Certificate2 certificate = _fixture.GetDefaultCertificate())
+            {
+                ArgumentNullException exception = Assert.Throws<ArgumentNullException>(
+                    () => DefaultX509ChainBuildPolicy.Instance.Build(chain: null, certificate));
+
+                Assert.Equal("chain", exception.ParamName);
+            }
+        }
+
+        [Fact]
+        public void Build_WhenCertificateIsNull_Throws()
+        {
+            using (var chain = new X509Chain())
+            {
+                ArgumentNullException exception = Assert.Throws<ArgumentNullException>(
+                    () => DefaultX509ChainBuildPolicy.Instance.Build(chain, certificate: null));
+
+                Assert.Equal("certificate", exception.ParamName);
+            }
+        }
+
+        [Fact]
+        public void Build_WhenArgumentsAreValid_ReturnsExpectedResult()
+        {
+            using (var chain = new X509Chain())
+            using (X509Certificate2 expectedCertificate = _fixture.GetDefaultCertificate())
+            {
+                bool actualResult = DefaultX509ChainBuildPolicy.Instance.Build(chain, expectedCertificate);
+
+                Assert.False(actualResult);
+                Assert.Equal(X509ChainStatusFlags.UntrustedRoot, chain.ChainStatus[0].Status);
+            }
+        }
+    }
+}

--- a/test/NuGet.Core.Tests/NuGet.Packaging.Test/SigningTests/ChainBuilding/RetriableX509ChainBuildPolicyTests.cs
+++ b/test/NuGet.Core.Tests/NuGet.Packaging.Test/SigningTests/ChainBuilding/RetriableX509ChainBuildPolicyTests.cs
@@ -2,8 +2,8 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
-using System.Diagnostics;
 using System.Security.Cryptography.X509Certificates;
+using Moq;
 using NuGet.Packaging.Signing;
 using NuGet.Test.Utility;
 using Xunit;
@@ -20,21 +20,36 @@ namespace NuGet.Packaging.Test
         }
 
         [PlatformFact(Platform.Windows)]
+        public void Constructor_WhenInnerPolicyIsNull_Throws()
+        {
+            ArgumentNullException exception = Assert.Throws<ArgumentNullException>(
+                () => new RetriableX509ChainBuildPolicy(innerPolicy: null, retryCount: 0, TimeSpan.MaxValue));
+
+            Assert.Equal("innerPolicy", exception.ParamName);
+        }
+
+        [PlatformFact(Platform.Windows)]
         public void Constructor_WhenRetryCountIsInvalid_Throws()
         {
+            var innerPolicy = new Mock<IX509ChainBuildPolicy>(MockBehavior.Strict);
             ArgumentOutOfRangeException exception = Assert.Throws<ArgumentOutOfRangeException>(
-                () => new RetriableX509ChainBuildPolicy(retryCount: 0, TimeSpan.MaxValue));
+                () => new RetriableX509ChainBuildPolicy(innerPolicy.Object, retryCount: 0, TimeSpan.MaxValue));
 
             Assert.Equal("retryCount", exception.ParamName);
+
+            innerPolicy.VerifyAll();
         }
 
         [PlatformFact(Platform.Windows)]
         public void Constructor_WhenSleepIntervalIsInvalid_Throws()
         {
+            var innerPolicy = new Mock<IX509ChainBuildPolicy>(MockBehavior.Strict);
             ArgumentOutOfRangeException exception = Assert.Throws<ArgumentOutOfRangeException>(
-                () => new RetriableX509ChainBuildPolicy(retryCount: 1, TimeSpan.FromSeconds(-1)));
+                () => new RetriableX509ChainBuildPolicy(innerPolicy.Object, retryCount: 1, TimeSpan.FromSeconds(-1)));
 
             Assert.Equal("sleepInterval", exception.ParamName);
+
+            innerPolicy.VerifyAll();
         }
 
         [PlatformTheory(Platform.Windows)]
@@ -44,18 +59,23 @@ namespace NuGet.Packaging.Test
             int expectedRetryCount,
             int expectedMilliseconds)
         {
+            var innerPolicy = new Mock<IX509ChainBuildPolicy>(MockBehavior.Strict);
             TimeSpan expectedSleepInterval = TimeSpan.FromSeconds(expectedMilliseconds);
 
-            var policy = new RetriableX509ChainBuildPolicy(expectedRetryCount, expectedSleepInterval);
+            var policy = new RetriableX509ChainBuildPolicy(innerPolicy.Object, expectedRetryCount, expectedSleepInterval);
 
+            Assert.Same(innerPolicy.Object, policy.InnerPolicy);
             Assert.Equal(expectedRetryCount, policy.RetryCount);
             Assert.Equal(expectedSleepInterval, policy.SleepInterval);
+
+            innerPolicy.VerifyAll();
         }
 
         [PlatformFact(Platform.Windows)]
         public void Build_WhenBuildIsNull_Throws()
         {
-            var policy = new RetriableX509ChainBuildPolicy(retryCount: 1, TimeSpan.Zero);
+            var innerPolicy = new Mock<IX509ChainBuildPolicy>(MockBehavior.Strict);
+            var policy = new RetriableX509ChainBuildPolicy(innerPolicy.Object, retryCount: 1, TimeSpan.Zero);
 
             using (X509Certificate2 certificate = _fixture.GetDefaultCertificate())
             {
@@ -63,13 +83,16 @@ namespace NuGet.Packaging.Test
                     () => policy.Build(chain: null, certificate));
 
                 Assert.Equal("chain", exception.ParamName);
+
+                innerPolicy.VerifyAll();
             }
         }
 
         [PlatformFact(Platform.Windows)]
         public void Create_WhenCertificateIsNull_Throws()
         {
-            var policy = new RetriableX509ChainBuildPolicy(retryCount: 1, TimeSpan.Zero);
+            var innerPolicy = new Mock<IX509ChainBuildPolicy>(MockBehavior.Strict);
+            var policy = new RetriableX509ChainBuildPolicy(innerPolicy.Object, retryCount: 1, TimeSpan.Zero);
 
             using (var chain = new X509Chain())
             {
@@ -77,57 +100,64 @@ namespace NuGet.Packaging.Test
                     () => policy.Build(chain, certificate: null));
 
                 Assert.Equal("certificate", exception.ParamName);
+
+                innerPolicy.VerifyAll();
             }
         }
 
         [PlatformFact(Platform.Windows)]
         public void Build_WhenArgumentsAreValidAndBuildSucceeds_DoesNotRetry()
         {
-            var policy = new RetriableX509ChainBuildPolicy(retryCount: 3, TimeSpan.FromSeconds(10));
+            var innerPolicy = new Mock<IX509ChainBuildPolicy>(MockBehavior.Strict);
+            var policy = new RetriableX509ChainBuildPolicy(innerPolicy.Object, retryCount: 3, TimeSpan.FromSeconds(10));
 
             using (var chain = new X509Chain())
-            using (X509Certificate2 certificate = GetTrustedCertificate())
+            using (X509Certificate2 certificate = _fixture.GetDefaultCertificate())
             {
-                Stopwatch stopwatch = Stopwatch.StartNew();
-
+                innerPolicy.Setup(
+                    x => x.Build(
+                        It.Is<X509Chain>(chainArg => ReferenceEquals(chainArg, chain)),
+                        It.Is<X509Certificate2>(certArg => ReferenceEquals(certArg, certificate))))
+                    .Returns(true);
                 bool actualResult = policy.Build(chain, certificate);
 
                 Assert.True(actualResult);
-                Assert.True(stopwatch.Elapsed < policy.SleepInterval);
+
+                innerPolicy.VerifyAll();
             }
         }
 
         [PlatformFact(Platform.Windows)]
         public void Build_WhenArgumentsAreValidAndBuildFailureIsRetriable_Retries()
         {
-            var policy = new RetriableX509ChainBuildPolicy(retryCount: 3, TimeSpan.FromMilliseconds(50));
+            var innerPolicy = new Mock<IX509ChainBuildPolicy>(MockBehavior.Strict);
+            var policy = new RetriableX509ChainBuildPolicy(innerPolicy.Object, retryCount: 3, TimeSpan.FromMilliseconds(50));
 
             using (var chain = new X509Chain())
             using (X509Certificate2 certificate = _fixture.GetDefaultCertificate())
             {
-                Stopwatch stopwatch = Stopwatch.StartNew();
+                innerPolicy.Setup(
+                   x => x.Build(
+                       It.Is<X509Chain>(chainArg => ReferenceEquals(chainArg, chain)),
+                       It.Is<X509Certificate2>(certArg => ReferenceEquals(certArg, certificate))))
+                   .Returns(false);
+                innerPolicy.Setup(
+                   x => x.Build(
+                       It.Is<X509Chain>(chainArg => ReferenceEquals(chainArg, chain)),
+                       It.Is<X509Certificate2>(certArg => ReferenceEquals(certArg, certificate))))
+                   .Returns(false);
+                innerPolicy.Setup(
+                    x => x.Build(
+                        It.Is<X509Chain>(chainArg => ReferenceEquals(chainArg, chain)),
+                        It.Is<X509Certificate2>(certArg => ReferenceEquals(certArg, certificate))))
+                    .Returns(true);
 
                 bool actualResult = policy.Build(chain, certificate);
 
-                Assert.False(actualResult);
-                Assert.True(stopwatch.Elapsed > TimeSpan.FromMilliseconds(policy.RetryCount * policy.SleepInterval.TotalMilliseconds));
+                Assert.True(actualResult);
+
+                innerPolicy.VerifyAll();
             }
-        }
-
-        private static X509Certificate2 GetTrustedCertificate()
-        {
-            X509Certificate2 certificate;
-
-            using (var store = new X509Store(StoreName.Root))
-            {
-                store.Open(OpenFlags.OpenExistingOnly | OpenFlags.ReadOnly);
-
-                certificate = store.Certificates[0];
-
-                store.Close();
-            }
-
-            return certificate;
         }
     }
 }

--- a/test/NuGet.Core.Tests/NuGet.Packaging.Test/SigningTests/ChainBuilding/RetriableX509ChainBuildPolicyTests.cs
+++ b/test/NuGet.Core.Tests/NuGet.Packaging.Test/SigningTests/ChainBuilding/RetriableX509ChainBuildPolicyTests.cs
@@ -1,0 +1,133 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.Security.Cryptography.X509Certificates;
+using NuGet.Packaging.Signing;
+using NuGet.Test.Utility;
+using Xunit;
+
+namespace NuGet.Packaging.Test
+{
+    public class RetriableX509ChainBuildPolicyTests : IClassFixture<CertificatesFixture>
+    {
+        private readonly CertificatesFixture _fixture;
+
+        public RetriableX509ChainBuildPolicyTests(CertificatesFixture fixture)
+        {
+            _fixture = fixture ?? throw new ArgumentNullException(nameof(fixture));
+        }
+
+        [PlatformFact(Platform.Windows)]
+        public void Constructor_WhenRetryCountIsInvalid_Throws()
+        {
+            ArgumentOutOfRangeException exception = Assert.Throws<ArgumentOutOfRangeException>(
+                () => new RetriableX509ChainBuildPolicy(retryCount: 0, TimeSpan.MaxValue));
+
+            Assert.Equal("retryCount", exception.ParamName);
+        }
+
+        [PlatformFact(Platform.Windows)]
+        public void Constructor_WhenSleepIntervalIsInvalid_Throws()
+        {
+            ArgumentOutOfRangeException exception = Assert.Throws<ArgumentOutOfRangeException>(
+                () => new RetriableX509ChainBuildPolicy(retryCount: 1, TimeSpan.FromSeconds(-1)));
+
+            Assert.Equal("sleepInterval", exception.ParamName);
+        }
+
+        [PlatformTheory(Platform.Windows)]
+        [InlineData(1, 0)]
+        [InlineData(3, 700)]
+        public void Constructor_WhenArgumentsAreValid_InitializesProperties(
+            int expectedRetryCount,
+            int expectedMilliseconds)
+        {
+            TimeSpan expectedSleepInterval = TimeSpan.FromSeconds(expectedMilliseconds);
+
+            var policy = new RetriableX509ChainBuildPolicy(expectedRetryCount, expectedSleepInterval);
+
+            Assert.Equal(expectedRetryCount, policy.RetryCount);
+            Assert.Equal(expectedSleepInterval, policy.SleepInterval);
+        }
+
+        [PlatformFact(Platform.Windows)]
+        public void Build_WhenBuildIsNull_Throws()
+        {
+            var policy = new RetriableX509ChainBuildPolicy(retryCount: 1, TimeSpan.Zero);
+
+            using (X509Certificate2 certificate = _fixture.GetDefaultCertificate())
+            {
+                ArgumentNullException exception = Assert.Throws<ArgumentNullException>(
+                    () => policy.Build(chain: null, certificate));
+
+                Assert.Equal("chain", exception.ParamName);
+            }
+        }
+
+        [PlatformFact(Platform.Windows)]
+        public void Create_WhenCertificateIsNull_Throws()
+        {
+            var policy = new RetriableX509ChainBuildPolicy(retryCount: 1, TimeSpan.Zero);
+
+            using (var chain = new X509Chain())
+            {
+                ArgumentNullException exception = Assert.Throws<ArgumentNullException>(
+                    () => policy.Build(chain, certificate: null));
+
+                Assert.Equal("certificate", exception.ParamName);
+            }
+        }
+
+        [PlatformFact(Platform.Windows)]
+        public void Build_WhenArgumentsAreValidAndBuildSucceeds_DoesNotRetry()
+        {
+            var policy = new RetriableX509ChainBuildPolicy(retryCount: 3, TimeSpan.FromSeconds(10));
+
+            using (var chain = new X509Chain())
+            using (X509Certificate2 certificate = GetTrustedCertificate())
+            {
+                Stopwatch stopwatch = Stopwatch.StartNew();
+
+                bool actualResult = policy.Build(chain, certificate);
+
+                Assert.True(actualResult);
+                Assert.True(stopwatch.Elapsed < policy.SleepInterval);
+            }
+        }
+
+        [PlatformFact(Platform.Windows)]
+        public void Build_WhenArgumentsAreValidAndBuildFailureIsRetriable_Retries()
+        {
+            var policy = new RetriableX509ChainBuildPolicy(retryCount: 3, TimeSpan.FromMilliseconds(50));
+
+            using (var chain = new X509Chain())
+            using (X509Certificate2 certificate = _fixture.GetDefaultCertificate())
+            {
+                Stopwatch stopwatch = Stopwatch.StartNew();
+
+                bool actualResult = policy.Build(chain, certificate);
+
+                Assert.False(actualResult);
+                Assert.True(stopwatch.Elapsed > TimeSpan.FromMilliseconds(policy.RetryCount * policy.SleepInterval.TotalMilliseconds));
+            }
+        }
+
+        private static X509Certificate2 GetTrustedCertificate()
+        {
+            X509Certificate2 certificate;
+
+            using (var store = new X509Store(StoreName.Root))
+            {
+                store.Open(OpenFlags.OpenExistingOnly | OpenFlags.ReadOnly);
+
+                certificate = store.Certificates[0];
+
+                store.Close();
+            }
+
+            return certificate;
+        }
+    }
+}

--- a/test/NuGet.Core.Tests/NuGet.Packaging.Test/SigningTests/ChainBuilding/X509ChainBuildPolicyFactoryTests.cs
+++ b/test/NuGet.Core.Tests/NuGet.Packaging.Test/SigningTests/ChainBuilding/X509ChainBuildPolicyFactoryTests.cs
@@ -21,6 +21,15 @@ namespace NuGet.Packaging.Test
             Assert.Equal("reader", exception.ParamName);
         }
 
+        [Fact]
+        public void Create_WhenArgumentIsValid_IsIdempotent()
+        {
+            IX509ChainBuildPolicy policy0 = X509ChainBuildPolicyFactory.Create(EnvironmentVariableWrapper.Instance);
+            IX509ChainBuildPolicy policy1 = X509ChainBuildPolicyFactory.Create(EnvironmentVariableWrapper.Instance);
+
+            Assert.Same(policy0, policy1);
+        }
+
         [PlatformTheory(Platform.Windows)]
         [InlineData(null)]
         [InlineData("")]
@@ -31,11 +40,11 @@ namespace NuGet.Packaging.Test
         [InlineData("0,1")]
         [InlineData("1,-2")]
         [InlineData("1,2,3")]
-        public void Create_OnWindowsWhenEnvironmentVariableIsInvalid_ReturnsDefaultPolicy(string value)
+        public void CreateWithoutCaching_OnWindowsWhenEnvironmentVariableIsInvalid_ReturnsDefaultPolicy(string value)
         {
             Mock<IEnvironmentVariableReader> reader = CreateMockEnvironmentVariableReader(value);
 
-            IX509ChainBuildPolicy policy = X509ChainBuildPolicyFactory.Create(reader.Object);
+            IX509ChainBuildPolicy policy = X509ChainBuildPolicyFactory.CreateWithoutCaching(reader.Object);
 
             Assert.IsType<DefaultX509ChainBuildPolicy>(policy);
 
@@ -45,11 +54,12 @@ namespace NuGet.Packaging.Test
         [PlatformTheory(Platform.Windows)]
         [InlineData("1,0")]
         [InlineData("3,7")]
-        public void Create_OnWindowsWhenEnvironmentVariableIsValid_ReturnsRetriablePolicy(string value)
+        [InlineData(" 5 , 9 ")]
+        public void CreateWithoutCaching_OnWindowsWhenEnvironmentVariableIsValid_ReturnsRetriablePolicy(string value)
         {
             Mock<IEnvironmentVariableReader> reader = CreateMockEnvironmentVariableReader(value);
 
-            IX509ChainBuildPolicy policy = X509ChainBuildPolicyFactory.Create(reader.Object);
+            IX509ChainBuildPolicy policy = X509ChainBuildPolicyFactory.CreateWithoutCaching(reader.Object);
 
             Assert.IsType<RetriableX509ChainBuildPolicy>(policy);
 
@@ -66,11 +76,11 @@ namespace NuGet.Packaging.Test
         }
 
         [PlatformFact(Platform.Linux, Platform.Darwin)]
-        public void Create_OnNonWindowsAlways_ReturnsDefaultPolicy()
+        public void CreateWithoutCaching_OnNonWindowsAlways_ReturnsDefaultPolicy()
         {
-            Mock<IEnvironmentVariableReader> reader = new Mock<IEnvironmentVariableReader>(MockBehavior.Strict);
+            var reader = new Mock<IEnvironmentVariableReader>(MockBehavior.Strict);
 
-            IX509ChainBuildPolicy policy = X509ChainBuildPolicyFactory.Create(reader.Object);
+            IX509ChainBuildPolicy policy = X509ChainBuildPolicyFactory.CreateWithoutCaching(reader.Object);
 
             Assert.IsType<DefaultX509ChainBuildPolicy>(policy);
 

--- a/test/NuGet.Core.Tests/NuGet.Packaging.Test/SigningTests/ChainBuilding/X509ChainBuildPolicyFactoryTests.cs
+++ b/test/NuGet.Core.Tests/NuGet.Packaging.Test/SigningTests/ChainBuilding/X509ChainBuildPolicyFactoryTests.cs
@@ -1,0 +1,91 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Moq;
+using NuGet.Common;
+using NuGet.Packaging.Signing;
+using NuGet.Test.Utility;
+using Xunit;
+
+namespace NuGet.Packaging.Test
+{
+    public class X509ChainBuildPolicyFactoryTests
+    {
+        [Fact]
+        public void Create_WhenArgumentIsNull_Throws()
+        {
+            ArgumentNullException exception = Assert.Throws<ArgumentNullException>(
+                () => X509ChainBuildPolicyFactory.Create(reader: null));
+
+            Assert.Equal("reader", exception.ParamName);
+        }
+
+        [PlatformTheory(Platform.Windows)]
+        [InlineData(null)]
+        [InlineData("")]
+        [InlineData(" ")]
+        [InlineData(",")]
+        [InlineData("-1,2")]
+        [InlineData("0,0")]
+        [InlineData("0,1")]
+        [InlineData("1,-2")]
+        [InlineData("1,2,3")]
+        public void Create_OnWindowsWhenEnvironmentVariableIsInvalid_ReturnsDefaultPolicy(string value)
+        {
+            Mock<IEnvironmentVariableReader> reader = CreateMockEnvironmentVariableReader(value);
+
+            IX509ChainBuildPolicy policy = X509ChainBuildPolicyFactory.Create(reader.Object);
+
+            Assert.IsType<DefaultX509ChainBuildPolicy>(policy);
+
+            reader.VerifyAll();
+        }
+
+        [PlatformTheory(Platform.Windows)]
+        [InlineData("1,0")]
+        [InlineData("3,7")]
+        public void Create_OnWindowsWhenEnvironmentVariableIsValid_ReturnsRetriablePolicy(string value)
+        {
+            Mock<IEnvironmentVariableReader> reader = CreateMockEnvironmentVariableReader(value);
+
+            IX509ChainBuildPolicy policy = X509ChainBuildPolicyFactory.Create(reader.Object);
+
+            Assert.IsType<RetriableX509ChainBuildPolicy>(policy);
+
+            var retryPolicy = (RetriableX509ChainBuildPolicy)policy;
+
+            string[] parts = value.Split(X509ChainBuildPolicyFactory.ValueDelimiter);
+            int expectedRetryCount = int.Parse(parts[0]);
+            TimeSpan expectedSleepInterval = TimeSpan.FromMilliseconds(int.Parse(parts[1]));
+
+            Assert.Equal(expectedRetryCount, retryPolicy.RetryCount);
+            Assert.Equal(expectedSleepInterval, retryPolicy.SleepInterval);
+
+            reader.VerifyAll();
+        }
+
+        [PlatformFact(Platform.Linux, Platform.Darwin)]
+        public void Create_OnNonWindowsAlways_ReturnsDefaultPolicy()
+        {
+            Mock<IEnvironmentVariableReader> reader = new Mock<IEnvironmentVariableReader>(MockBehavior.Strict);
+
+            IX509ChainBuildPolicy policy = X509ChainBuildPolicyFactory.Create(reader.Object);
+
+            Assert.IsType<DefaultX509ChainBuildPolicy>(policy);
+
+            reader.VerifyAll();
+        }
+
+        private static Mock<IEnvironmentVariableReader> CreateMockEnvironmentVariableReader(
+            string variableValue)
+        {
+            var reader = new Mock<IEnvironmentVariableReader>(MockBehavior.Strict);
+
+            reader.Setup(r => r.GetEnvironmentVariable(X509ChainBuildPolicyFactory.EnvironmentVariableName))
+                .Returns(variableValue);
+
+            return reader;
+        }
+    }
+}

